In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Load data

In [4]:
U = np.load('U.npy')
U

array([[-0.42344517,  0.55665609, -0.24102467,  0.3305458 , -0.24562245],
       [-0.50699337, -0.47790996, -0.16495257,  0.37953185,  0.58331032],
       [-0.37335489,  0.50609813, -0.19182881, -0.4734306 ,  0.385749  ],
       [-0.34447683,  0.12820357,  0.63388605,  0.41223622, -0.23721861],
       [-0.42422527, -0.40791144, -0.38601146, -0.2525886 , -0.62705737],
       [-0.35443853, -0.15090679,  0.57191179, -0.53741866,  0.033805  ]])

In [5]:
S = np.load('S.npy')
S

array([13.2663997 ,  7.31094493,  4.68971622,  0.89013458,  0.51666749])

In [6]:
V = np.load('V.npy')
V

array([[-0.35720733, -0.41849995, -0.66295151, -0.44991101, -0.23523627],
       [ 0.58989124,  0.53832973, -0.23397095, -0.4730836 , -0.2892731 ],
       [-0.16497259,  0.05617741,  0.14276305,  0.34480327, -0.91123957],
       [ 0.41157588, -0.10910113, -0.62130363,  0.65416345,  0.06895084],
       [ 0.57255619, -0.72111021,  0.31520375, -0.16417331, -0.16085148]])

# (a)

In [7]:
U_user1 = U[0,:]
U_user1

array([-0.42344517,  0.55665609, -0.24102467,  0.3305458 , -0.24562245])

In [8]:
V_Amelie = np.transpose(V[:,3])
V_Amelie

array([-0.44991101, -0.4730836 ,  0.34480327,  0.65416345, -0.16417331])

In [9]:
S_diag = np.diag(S)
S_diag

array([[13.2663997 ,  0.        ,  0.        ,  0.        ,  0.        ],
       [ 0.        ,  7.31094493,  0.        ,  0.        ,  0.        ],
       [ 0.        ,  0.        ,  4.68971622,  0.        ,  0.        ],
       [ 0.        ,  0.        ,  0.        ,  0.89013458,  0.        ],
       [ 0.        ,  0.        ,  0.        ,  0.        ,  0.51666749]])

In [21]:
#R_User1 = U_user1 @ S_diag @ V_Amelie
R_User1 =np.dot(np.dot(U_user1,S_diag),V_Amelie)
R_User1

0.4256821525055139

So user 1’s ‘estimated’ rating of the movie Amelie is 0.4256821525055139.

# (b)

The “strength” of the concept is given by the magnitude of the values in the SVD diagonal matrix.
And if you arrange it in decreasing order of the strength, the first element has the highest strength.

The strength will be 13.2663997.

# (c)

In [11]:
V_Movie3 = V[:,2]
V_Movie3

array([-0.66295151, -0.23397095,  0.14276305, -0.62130363,  0.31520375])

In [12]:
R_Movie3 = U @ S_diag @ V_Movie3
R_Movie3

array([2.38782659, 5.05114346, 2.6141595 , 2.96814942, 4.20793614,
       4.06103988])

In [13]:
# actual ratings of user2, 4, 6 exist
R_average3 = (2.38782659 + 2.5 + 2.6141595 + 4.5 + 4.20793614 + 4.5)/6
R_average3

3.451653705

So the average rating for Movie 3 across all users in the system is 3.5483758311270606

# (d)

In [14]:
R_all = U @ S_diag @ V
R_all

array([[ 4.64222751,  4.53770548,  2.38782659,  0.42568215,  1.21492044],
       [ 0.78072353,  0.6362689 ,  5.05114346,  4.5838201 ,  3.27264048],
       [ 4.04098583,  3.91643009,  2.6141595 , -0.14058136,  0.85347279],
       [ 1.77575468,  2.63245029,  2.96814942,  2.8978389 , -1.85997836],
       [ 0.27177039,  0.90633516,  4.20793614,  3.22482782,  3.87278406],
       [ 0.39946186,  1.5641847 ,  4.06103988,  3.24647193, -1.05457648]])

In [15]:
R_average1 = (5 + 0.78072353 + 4 + 1.77575468 + 1 + 0.39946186)/6
R_average1

2.1593233450000002

In [16]:
R_average2 = (4.5 + 0.6362689  + 4.5 + 1.5 + 0.90633516 + 1.5641847)/6
R_average2

2.267798126666667

In [17]:
R_average3

3.451653705

In [18]:
R_average4 = (0.42568215 + 5  -0.14058136 + 2.8978389 + 0 + 3.24647193)/6
R_average4

1.9049019366666669

In [19]:
R_average5 = (1 + 5  + 0.85347279 -1.85997836 + 5 -1.05457648)/6
R_average5

1.4898196583333334

The movie with the overall highest rating is Harry Potter